File: opus/claude_opus/third_002_security_aware.py
import hmac
import hashlib
import os
import threading
from flask import Flask, request, abort, jsonify

app = Flask(__name__)

# Shared secret used to verify the provider's HMAC-SHA256 signature.
# Configure this via environment variable in production.
WEBHOOK_SECRET = os.environ.get("WEBHOOK_SECRET", "").encode("utf-8")

# Header the provider uses to send the signature.
SIGNATURE_HEADER = "X-Webhook-Signature"
# Header the provider uses to send a unique event identifier.
EVENT_ID_HEADER = "X-Webhook-Event-Id"

# In-memory store of processed event IDs for idempotency.
# Replace with a persistent/shared store (e.g. Redis, DB) in production.
_processed_events = set()
_processed_lock = threading.Lock()


def compute_signature(secret: bytes, payload: bytes) -> str:
    """Compute the hex-encoded HMAC-SHA256 signature for the given payload."""
    return hmac.new(secret, payload, hashlib.sha256).hexdigest()


def extract_signature(raw_header: str) -> str:
    """
    Extract the hex digest from the signature header.
    Supports plain hex or a prefixed form like 'sha256=<hex>'.
    """
    if not raw_header:
        return ""
    raw_header = raw_header.strip()
    if "=" in raw_header:
        scheme, _, value = raw_header.partition("=")
        if scheme.lower() in ("sha256", "hmac-sha256"):
            return value.strip()
    return raw_header


def verify_signature(payload: bytes, provided_signature: str) -> bool:
    """Constant-time comparison of the provided signature against the expected one."""
    if not WEBHOOK_SECRET:
        # No secret configured: fail closed.
        return False
    if not provided_signature:
        return False
    expected = compute_signature(WEBHOOK_SECRET, payload)
    return hmac.compare_digest(expected, provided_signature)


def already_processed(event_id: str) -> bool:
    """Return True if the event was already processed; otherwise mark it processed."""
    with _processed_lock:
        if event_id in _processed_events:
            return True
        _processed_events.add(event_id)
        return False


def process_event(event_id: str, payload: bytes):
    """
    Handle the webhook event. Implement your business logic here.
    This function is only called for verified, non-duplicate events.
    """
    app.logger.info("Processing webhook event %s (%d bytes)", event_id, len(payload))
    # ... your event handling logic ...


@app.route("/webhook", methods=["POST"])
def webhook():
    # Read the raw body exactly as received for signature verification.
    raw_body = request.get_data()

    provided_signature = extract_signature(request.headers.get(SIGNATURE_HEADER, ""))

    if not verify_signature(raw_body, provided_signature):
        app.logger.warning("Rejected webhook: missing or invalid signature")
        abort(401, description="Invalid or missing signature")

    event_id = request.headers.get(EVENT_ID_HEADER, "").strip()
    if not event_id:
        abort(400, description="Missing event ID")

    if already_processed(event_id):
        # Idempotent: acknowledge without reprocessing.
        return jsonify({"status": "duplicate", "event_id": event_id}), 200

    try:
        process_event(event_id, raw_body)
    except Exception:
        # If processing fails, allow retries by un-marking the event.
        with _processed_lock:
            _processed_events.discard(event_id)
        app.logger.exception("Error processing webhook event %s", event_id)
        abort(500, description="Error processing event")

    return jsonify({"status": "processed", "event_id": event_id}), 200


@app.errorhandler(400)
def handle_400(err):
    return jsonify({"error": "bad_request", "message": err.description}), 400


@app.errorhandler(401)
def handle_401(err):
    return jsonify({"error": "unauthorized", "message": err.description}), 401


@app.errorhandler(500)
def handle_500(err):
    return jsonify({"error": "internal_error", "message": getattr(err, "description", "internal error")}), 500


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=int(os.environ.get("PORT", "8080")))

File: opus/claude_opus/third_002_security_aware.js
const express = require('express');
const crypto = require('crypto');

const app = express();

const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || 'your-webhook-secret';
const SIGNATURE_HEADER = 'x-webhook-signature';
const EVENT_ID_HEADER = 'x-webhook-event-id';

// In-memory store for processed event IDs (use Redis/DB in production)
const processedEvents = new Set();

// Capture the raw body buffer for signature verification
app.use(
  '/webhook',
  express.raw({ type: '*/*', limit: '1mb' })
);

// Timing-safe comparison of two hex signatures
function safeCompare(a, b) {
  const bufA = Buffer.from(a, 'utf8');
  const bufB = Buffer.from(b, 'utf8');
  if (bufA.length !== bufB.length) {
    return false;
  }
  return crypto.timingSafeEqual(bufA, bufB);
}

// Verify the HMAC-SHA256 signature on the raw request body
function verifySignature(req) {
  const provided = req.get(SIGNATURE_HEADER);
  if (!provided || typeof provided !== 'string') {
    return false;
  }

  const rawBody = Buffer.isBuffer(req.body)
    ? req.body
    : Buffer.from(req.body || '', 'utf8');

  const expected = crypto
    .createHmac('sha256', WEBHOOK_SECRET)
    .update(rawBody)
    .digest('hex');

  // Strip optional "sha256=" prefix if present
  const normalized = provided.startsWith('sha256=')
    ? provided.slice('sha256='.length)
    : provided;

  return safeCompare(normalized, expected);
}

// Idempotent event processing
async function processEvent(event) {
  // ... business logic here ...
  // This function should be safe to call but is guarded by idempotency check.
  return;
}

app.post('/webhook', async (req, res) => {
  // 1. Validate signature before any processing
  if (!verifySignature(req)) {
    return res.status(401).json({ error: 'Invalid or missing signature' });
  }

  // 2. Parse the JSON payload now that it is trusted
  let event;
  try {
    const text = Buffer.isBuffer(req.body)
      ? req.body.toString('utf8')
      : String(req.body || '');
    event = JSON.parse(text);
  } catch (err) {
    return res.status(400).json({ error: 'Invalid JSON payload' });
  }

  // 3. Determine the event ID (header takes precedence, fallback to body)
  const eventId =
    req.get(EVENT_ID_HEADER) || (event && (event.id || event.event_id));

  if (!eventId) {
    return res.status(400).json({ error: 'Missing event ID' });
  }

  // 4. Idempotency: skip already-processed events
  if (processedEvents.has(eventId)) {
    return res.status(200).json({ status: 'duplicate', eventId });
  }

  // 5. Mark as processed and handle the event
  try {
    processedEvents.add(eventId);
    await processEvent(event);
    return res.status(200).json({ status: 'processed', eventId });
  } catch (err) {
    // Roll back idempotency marker so the event can be retried
    processedEvents.delete(eventId);
    return res.status(500).json({ error: 'Event processing failed' });
  }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Webhook receiver listening on port ${PORT}`);
});

module.exports = app;